### PR TITLE
support conda pack in init_spark_on_k8s

### PR DESCRIPTION
**Pull request by Le-Zheng**
standard spark-submit commond:
```
/opt/spark/bin/spark-submit \
  --master k8s://https://172.16.0.200:6443 \
  --deploy-mode client \
  --conf spark.driver.host=172.16.0.188 \
  --conf spark.driver.port=54321 \
  --conf spark.kubernetes.authenticate.driver.serviceAccountName=spark \
  --name test-bigdl2-pytorch_estimator \
  --conf spark.kubernetes.container.image=10.239.45.10/arda/intelanalytics/bigdl-k8s-spark-3.1.2:0.14.0-SNAPSHOT \
  --conf spark.executor.instances=1 \
  --conf spark.kubernetes.driver.volumes.persistentVolumeClaim.nfsvolumeclaim.options.claimName=nfsvolumeclaim \
  --conf spark.kubernetes.driver.volumes.persistentVolumeClaim.nfsvolumeclaim.mount.path=/bigdl2.0/data \
  --conf spark.kubernetes.executor.volumes.persistentVolumeClaim.nfsvolumeclaim.options.claimName=nfsvolumeclaim \
  --conf spark.kubernetes.executor.volumes.persistentVolumeClaim.nfsvolumeclaim.mount.path=/bigdl2.0/data \
  --conf spark.kubernetes.executor.deleteOnTermination=false \
  --conf spark.kubernetes.driver.label.az=true \
  --conf spark.kubernetes.executor.label.az=true \
  --conf spark.kubernetes.node.selector.spark=true \
  --conf spark.kubernetes.driverEnv.http_proxy=http://child-prc.intel.com:913 \
  --conf spark.kubernetes.driverEnv.https_proxy=http://child-prc.intel.com:913 \
  --conf spark.kubernetes.executorEnv.http_proxy=http://child-prc.intel.com:913 \
  --conf spark.kubernetes.executorEnv.https_proxy=http://child-prc.intel.com:913 \
  --conf spark.kubernetes.container.image.pullPolicy=Always \
  --conf spark.pyspark.driver.python=python \
  --conf spark.pyspark.python=./env/bin/python \
  --conf spark.kubernetes.executor.deleteOnTermination=false \
  --archives ./environment.tar.gz#env \
  --executor-cores 16 \
  --executor-memory 50g \
  --total-executor-cores 64 \
  --driver-cores 4 \
  --driver-memory 50g \
  --properties-file /opt/bigdl-0.14.0-SNAPSHOT/conf/spark-bigdl.conf \
  --py-files /opt/bigdl-0.14.0-SNAPSHOT/python/bigdl-spark_3.1.2-0.14.0-SNAPSHOT-python-api.zip,/opt/bigdl-0.14.0-SNAPSHOT/examples/orca/learn/horovod/pytorch_estimator.py \
  --conf spark.driver.extraJavaOptions=-Dderby.stream.error.file=/tmp \
  --conf spark.sql.catalogImplementation='in-memory' \
  --conf spark.driver.extraClassPath=local:///opt/bigdl-0.14.0-SNAPSHOT/jars/* \
  --conf spark.executor.extraClassPath=local:///opt/bigdl-0.14.0-SNAPSHOT/jars/* \
  local:///opt/bigdl-0.14.0-SNAPSHOT/examples/orca/learn/horovod/pytorch_estimator.py \
  --cluster_mode 'spark-submit'
```

### CODE REVIEW

**Comment by hkvision** [comment]
Will --archive automatically distribute the conda env package to all pods like yarn?

**Comment by Le-Zheng** [comment]
> Will --archive automatically distribute the conda env package to all pods like yarn?

yes.

**Comment by hkvision** [comment]
> > Will --archive automatically distribute the conda env package to all pods like yarn?
> 
> yes.

So in this way shall we still have a heavy docker image? The python environment can be removed from the docker?

**Comment by Le-Zheng** [comment]
> > > Will --archive automatically distribute the conda env package to all pods like yarn?
> > 
> > 
> > yes.
> 
> So in this way shall we still have a heavy docker image? The python environment can be removed from the docker?

Yes. plan to remove python environment from the docker https://github.com/intel-analytics/BigDL/pull/3717

File: python/dllib/src/bigdl/dllib/utils/spark.py [hunk 1]
```diff
@@ -313,54 +313,70 @@ def stop_spark_standalone():
...
+                         "spark.executorEnv.LD_PRELOAD": preload_so,
+                         "spark.kubernetes.container.image": container_image})
+            if "spark.driver.host" not in conf:
+                conf["spark.driver.host"] = get_node_ip()
+            if "spark.driver.port" not in conf:
+                conf["spark.driver.port"] = random.randint(10000, 65535)
+            if "BIGDL_CLASSPATH" in os.environ:
+                zoo_bigdl_jar_path = os.environ["BIGDL_CLASSPATH"]
+            else:
+                zoo_bigdl_jar_path = get_zoo_bigdl_classpath_on_driver()
```
Comment (hkvision) [inline python/dllib/src/bigdl/dllib/utils/spark.py:369]: This is not correct, the classpath on driver is not the same as that on executor. Need to modify according to yarn and test some training example.

File: python/dllib/src/bigdl/dllib/nncontext.py [hunk 2]
```diff
@@ -275,27 +275,30 @@ def init_spark_standalone(num_executors,
...
                       container_image,
+                      conda_name,
                       num_executors,
                       executor_cores,
                       executor_memory="2g",
                       driver_memory="1g",
                       driver_cores=4,
                       extra_executor_memory_for_ray=None,
                       extra_python_lib=None,
+                      penv_archive=None,
```
Comment (hkvision) [inline python/dllib/src/bigdl/dllib/nncontext.py:286]: Shall we remove this argument for penv_archive for k8s and yarn as well?
As the executor path is hardcoded as python_env, the penv_archive provided by the user wouldn't follow this name and there will be error.

File: python/dllib/src/bigdl/dllib/utils/spark.py (hunk 1 above)
Comment (Le-Zheng) [reply python/dllib/src/bigdl/dllib/utils/spark.py]: You are right. 
Updated. zoo_bigdl_path_on_executor is `python_env/lib/python3.7/site-packages/bigdl/share/dllib/lib/bigdl-dllib-spark_3.1.2-0.14.0-SNAPSHOT-jar-with-dependencies.jar` now.

File: python/dllib/src/bigdl/dllib/nncontext.py (hunk 2 above)
Comment (Le-Zheng) [reply python/dllib/src/bigdl/dllib/nncontext.py:286]: It's not conflict. "penv_archive" is a python env .tar path. Either "penv_archive" or conda pack default env .tar will be specified as `--archives env.tar#python_env`.

Comment (hkvision) [reply python/dllib/src/bigdl/dllib/nncontext.py:286]: archive = "{}#{}".format(penv_archive, executor_python_env)
But how to guarantee the user specified penv_archieve has `python_env` as we expect?

**Review by hkvision — APPROVED** [review]

**Pull request merged**